libraries/AP_HAL_ChibiOS/hwdef/Aocoda-RC-H743Dual/hwdef.dat

```diff
@@ -0,0 +1,218 @@
+
+# OTG USB Detect
+PE2 EXT_CS2 CS
+
+# These are the pins for SWD debugging with a STlinkv2 or black-magic probe.
+PA13 JTMS-SWDIO SWD
+PA14 JTCK-SWCLK SWD
+
+# SPI1 for IMU1 (BMI270/MPU6000)
+PC15 IMU1_CS CS
+PA5 SPI1_SCK SPI1
+PA6 SPI1_MISO SPI1
+PD7 SPI1_MOSI SPI1
+
+# SPI2 for MAX7456 OSD
+PB12 MAX7456_CS CS
+PB13 SPI2_SCK SPI2
+PB14 SPI2_MISO SPI2
+PB15 SPI2_MOSI SPI2
+
+# SPI3 for SPI flash - W25N01GV
+PD3 FLASH_CS CS
+PB3 SPI3_SCK SPI3
+PB4 SPI3_MISO SPI3
+PB5 SPI3_MOSI SPI3
+
+# SPI4 for IMU2 (BMI270)
+PE11 IMU2_CS CS
+PE12 SPI4_SCK SPI4
+PE13 SPI4_MISO SPI4
+PE14 SPI4_MOSI SPI4
+
+# two I2C bus, order of I2C buses
+I2C_ORDER I2C2 I2C1
+
+# I2C1
+PB6 I2C1_SCL I2C1
+PB7 I2C1_SDA I2C1
+
+# I2C2
+PB10 I2C2_SCL I2C2
+PB11 I2C2_SDA I2C2
+
+# ADC for Power
+PC0 BATT_VOLTAGE_SENS ADC1 SCALE(1)
+PC1 BATT_CURRENT_SENS ADC1 SCALE(1)
+
+define HAL_BATT_MONITOR_DEFAULT 4
+define HAL_BATT_VOLT_PIN 10
+define HAL_BATT_CURR_PIN 11
+define HAL_BATT_VOLT_SCALE 11.0
+define HAL_BATT_CURR_SCALE 40.0
+
+# ADC for airspeed
+PC4 PRESSURE_SENS ADC1 SCALE(2)
+define HAL_DEFAULT_AIRSPEED_PIN 4
+
+# ADC for rssi
+PC5 RSSI_ADC ADC1
+define BOARD_RSSI_ANA_PIN 8
+
+# Define LEDs
+# blue LED0 marked as ACT
+# green LED1 marked as B/E
+PE3 LED0 OUTPUT LOW GPIO(90) # blue
+PE4 LED1 OUTPUT LOW GPIO(91) # orange
+define HAL_GPIO_A_LED_PIN 91
+define HAL_GPIO_B_LED_PIN 90
+define HAL_GPIO_LED_OFF 1
+
+# The normal usage of this ordering is:
+# 1) SERIAL0: console (primary mavlink, usually USB)
+# 2) SERIAL1: RC input
+# 3) SERIAL2: GPS
+# 4) SERIAL3: VTX
+# 5) SERIAL4: Telemetry
+# 6) SERIAL5: not available
+# 7) SERIAL6: ESC Telemetry
+# 8) SERIAL7: spare
+# 9) SERIAL8: Optical Flow
+
+# order of UARTs (and USB)
+SERIAL_ORDER OTG1 USART1 USART2 USART3 UART4 EMPTY USART6 UART7 UART8
+
+# USART1 (RC input), SERIAL1
+PA9  USART1_TX USART1
+PA10 TIM1_CH3 TIM1 RCININT PULLDOWN LOW
+
+# as an alternative config setup the RX1 pin as a uart. This allows
+# for bi-directional UART based receiver protocols such as FPort
+# without any extra hardware
+PA10 USART1_RX USART1 ALT(1)
+
+# USART2 (GPS), SERIAL2
+PD5 USART2_TX USART2
+PD6 USART2_RX USART2
+
+# USART3 (VTX), SERIAL3
+PD8 USART3_TX USART3
+PD9 USART3_RX USART3
+
+# UART4 (Telemetry), SERIAL4
+PB8 UART4_RX UART4
+PB9 UART4_TX UART4
+
+# USART6 (ESC Telemetry), SERIAL6
+PC6 USART6_TX USART6
+PC7 USART6_RX USART6
+
+# UART7 (spare), SERIA7
+PE7 UART7_RX UART7
+PE8 UART7_TX UART7
+
+# UART8 (Optical Flow), SERIAL8
+PE0 UART8_RX UART8
+PE1 UART8_TX UART8
+
+# Motors
+PB0  TIM3_CH3 TIM3  PWM(1)  GPIO(50) BIDIR   //S1
+PB1  TIM3_CH4 TIM3  PWM(2)  GPIO(51)         //S2
+PA0  TIM2_CH1 TIM2  PWM(3)  GPIO(52) BIDIR   //S3
+PA1  TIM2_CH2 TIM2  PWM(4)  GPIO(53)         //S4
+PA2  TIM5_CH3 TIM5  PWM(5)  GPIO(54) BIDIR   //S5
+PA3  TIM5_CH4 TIM5  PWM(6)  GPIO(55)         //S6
+PD12 TIM4_CH1 TIM4  PWM(7)  GPIO(56) BIDIR   //S7
+PD13 TIM4_CH2 TIM4  PWM(8)  GPIO(57)         //S8
+PD14 TIM4_CH3  TIM4  PWM(9)  GPIO(58) NODMA  //S9

```

Comment:
You can't do this if you want dshot on PWM(7) and PWM(8). You should take off the NODMA

Reply:
Is it because these pins are all PDx? And if I remove "NODMA", PWM(9) PWM(10) can also be use as bdshot, right?

change from 

```
PD12 TIM4_CH1 TIM4  PWM(7)  GPIO(56) BIDIR   //S7
PD13 TIM4_CH2 TIM4  PWM(8)  GPIO(57)         //S8
PD14 TIM4_CH3  TIM4  PWM(9)  GPIO(58) NODMA  //S9
PD15 TIM4_CH4  TIM4  PWM(10) GPIO(59) NODMA  //S10
```
to 

```
PD12 TIM4_CH1 TIM4  PWM(7)  GPIO(56) BIDIR   //S7
PD13 TIM4_CH2 TIM4  PWM(8)  GPIO(57)         //S8
PD14 TIM4_CH3  TIM4  PWM(9)  GPIO(58)        //S9
PD15 TIM4_CH4  TIM4  PWM(10) GPIO(59)        //S10
```

Reply:
To enable bi-dir on all you would need to add BIDIR to PWM(9) as well. More what I was saying was that dshot won't work at all on PWM7-10 without this change

Reply:
> Is it because these pins are all PDx

No, its because they are all on TIM4

Reply:
> To enable bi-dir on all you would need to add BIDIR to PWM(9) as well. 

OK.  I though PWM(1) to PWM(8) works for bdshot is OK, as there are only two motor connector wiring PWM(1) to PWM(8).

I'm NOT aware of how code handling bdshot. Is BIDIR lable working on TIM4_CHn and TIM4_CHn+1, n belongs to odd numbers?

> More what I was saying was that dshot won't work at all on PWM7-10 without this change

Thanks.

Reply:
BIDIR applies to channel pairs - either CH1-2 or CH3-4. It can be on any channel apart from TIM4_CH4 which does not exist.

Reply:
I don't understand about " It can be on any channel apart from TIM4_CH4 which does not exist."
As I saw TIM4_CH4 definition below.

![图片](https://github.com/ArduPilot/ardupilot/assets/1296171/17cbc26c-74aa-403c-a314-d827716e30ad)

![图片](https://github.com/ArduPilot/ardupilot/assets/1296171/1efdc613-8676-439e-8e87-639a71670323)

Reply:
Look at the DMA channel map - TIM4_CH4 is missing

Reply:
Yes, it's missing. Thanks.

![图片](https://github.com/ArduPilot/ardupilot/assets/1296171/7f21296f-fc63-4ab7-8470-c11f50dfefe7)

Reply:
Sorry, but you should add BIDIR to TIM4_CH3 - we don't currently have the ability to do split dshot/bdshot on the same timer

Reply:
Currently the layout has following suggestions:
a) S1-S8 output throught ESC connector, which is supposed to be used as bdshot by default.
b) S9 S10 used for servos
c) SE1, SE2 used for servos or pinio

So I have left these IO definitions below:

```

# Motors
PB0  TIM3_CH3 TIM3  PWM(1)  GPIO(50) BIDIR   //S1
PB1  TIM3_CH4 TIM3  PWM(2)  GPIO(51)         //S2
PA0  TIM2_CH1 TIM2  PWM(3)  GPIO(52) BIDIR   //S3
PA1  TIM2_CH2 TIM2  PWM(4)  GPIO(53)         //S4
PA2  TIM5_CH3 TIM5  PWM(5)  GPIO(54) BIDIR   //S5
PA3  TIM5_CH4 TIM5  PWM(6)  GPIO(55)         //S6
PD12 TIM4_CH1 TIM4  PWM(7)  GPIO(56) BIDIR   //S7
PD13 TIM4_CH2 TIM4  PWM(8)  GPIO(57)         //S8
PD14 TIM4_CH3  TIM4  PWM(9)  GPIO(58)        //S9
PD15 TIM4_CH4  TIM4  PWM(10) GPIO(59)        //S10
PE5  TIM15_CH1 TIM15 PWM(11) GPIO(60) NODMA  //SE1
PE6  TIM15_CH2 TIM15 PWM(12) GPIO(61) NODMA  //SE2
```

Reply:
this stikk needs the change Andy mentioned to add BIDIR to PD14 line

Reply:
@Hwurzburg @andyp1per OK, I'll add BIDIR to PD14 line. But I don't understand why?

As andy and I have discussion about this and in his words "BIDIR applies to channel pairs - either CH1-2 or CH3-4. It can be on any channel apart from TIM4_CH4 which does not exist."

Now I just wana use S1-S8 left S9/S10 no bdshot (maybe saving DMA), why I have to add BIDIR to PD14?

One more question if I add BIDIR to PD14, does PD14/PD15 (S9/S10) can use as bdshot pins?